File: cpp/example7.cpp
#include <iostream>
#include <cstddef>
#include <cstdlib>
#include "exevalator.hpp"
#include "exevalator.cpp"

/**
 * An example to compute the numerical integration value of the inputted expression f(x).
 * For details of the numerical integration algorithm used in this code, see:
 *   https://www.vcssl.org/en-us/code/archive/0001/7800-vnano-integral-output/   (English)
 *   https://www.vcssl.org/ja-jp/code/archive/0001/7800-vnano-integral-output/   (Japanese)
 */
int main() {

    // Get the expression from the standard-input
    std::string expression;
    std::cout << "" << std::endl;
    std::cout << "This program computes the value of f(x) at x." << std::endl;
    std::cout << "" << std::endl;
    std::cout << "f(x) = ?               (example: 3*x*x)" << std::endl;
    std::cin >> expression;

    // Get the value of the lower limit from the standard-input
    double lower_limit;
    std::cout << "lower-limit = ?                  (example: 1)" << std::endl;
    std::cin >> lower_limit;

    // Get the value of the upper limit from the standard-input
    double upper_limit;
    std::cout << "upper-limit = ?                  (example: 2)" << std::endl;
    std::cin >> upper_limit;

    // Other numerical integration parameters
    long number_of_steps = 65536;
    double delta = (upper_limit - lower_limit) / number_of_steps;
    double result = 0.0;

    // Create an instance of Exevalator Engine
    exevalator::Exevalator exevalator;
    std::size_t x_address = exevalator.declare_variable("x");

    try {

        // Traverse tiny intervals from lower-limit to upper-limit
        for (long i=0; i<number_of_steps; ++i) {

            // The x-coordinate value of the left-bottom point of i-th tiny interval
            double x = lower_limit + i * delta;

            // Compute area of i-th tiny interval approximately by using Simpson's rule,
            // and add it to the variable "result"
            // (see: https://www.vcssl.org/en-us/code/archive/0001/7800-vnano-integral-output/ )

            exevalator.write_variable_at(x_address, x);
            double fx_left = exevalator.eval(expression);

            exevalator.write_variable_at(x_address, x + delta);
            double fx_right = exevalator.eval(expression);

            exevalator.write_variable_at(x_address, x + delta/2.0);
            double fx_center = exevalator.eval(expression);

            result += (fx_left + fx_right + 4.0 * fx_center) * delta / 6.0;
        }

        // Display the result
        std::cout << "----------" << std::endl;
        std::cout << "f(x)        = " << expression << std::endl;
        std::cout << "lower-limit = " << lower_limit << std::endl;
        std::cout << "upper-limit = " << upper_limit << std::endl;
        std::cout << "result      = " << result << std::endl;

    } catch (exevalator::ExevalatorException &e) {
        std::cout << "Error occurred: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: rust/example7.rs
mod exevalator;
use exevalator::Exevalator;
use std::io;

/// An example to compute the numerical integration value of the inputted expression f(x).
/// For details of the numerical integration algorithm used in this code, see:
///   https://www.vcssl.org/en-us/code/archive/0001/7800-vnano-integral-output/   (English)
///   https://www.vcssl.org/ja-jp/code/archive/0001/7800-vnano-integral-output/   (Japanese)
fn main() {

    println!("");
    println!("This program computes the value of f(x) at x.");
    println!("");

    // Get the expression from the standard-input
    println!("f(x) = ?               (default: 3*x*x + 2*x + 1)");
    let mut expression_string: String = String::new();
    io::stdin().read_line(&mut expression_string).expect("I/O error occurred for receiving the expression.");
    let mut expression_str: &str = expression_string.trim();
    if expression_str.is_empty() {
        expression_str = "3*x*x + 2*x + 1";
    }

    // Get the value of lower-limit from the standard-input
    println!("lower-limit = ?                  (default: 0)");
    let mut lower_limit_string: String = String::new();
    io::stdin().read_line(&mut lower_limit_string).expect("I/O error occurred for receiving the lower-limit value.");
    let mut lower_limit_str: &str = lower_limit_string.trim();
    if lower_limit_str.is_empty() {
        lower_limit_str = "0";
    }
    let lower_limit_value: f64 = lower_limit_str.parse::<f64>().expect("Invalid lower-limit value.");

    // Get the value of upper-limit from the standard-input
    println!("upper-limit = ?                  (default: 1)");
    let mut upper_limit_string: String = String::new();
    io::stdin().read_line(&mut upper_limit_string).expect("I/O error occurred for receiving the upper-limit value.");
    let mut upper_limit_str: &str = upper_limit_string.trim();
    if upper_limit_str.is_empty() {
        upper_limit_str = "1";
    }
    let upper_limit_value: f64 = upper_limit_str.parse::<f64>().expect("Invalid upper-limit value.");

    // Other numerical integration parameters
    let number_of_steps: i64 = 65536;
    let delta: f64 = (upper_limit_value - lower_limit_value) / (number_of_steps as f64);
    let mut result: f64 = 0.0;

    // Create an instance of Exevalator Engine
    let mut exevalator = Exevalator::new();
    let x_address = exevalator.declare_variable("x").expect("Failed to declare the variable x.");

    // Traverse tiny intervals from lower-limit to upper-limit
    for i in 0..number_of_steps {

        // The x-coordinate value of the left-bottom point of i-th tiny interval
        let x: f64 = lower_limit_value + (i as f64) * delta;

        // Compute area of i-th tiny interval approximately by using Simpson's rule,
        // and add it to the variable "result"
        // (see: https://www.vcssl.org/en-us/code/archive/0001/7800-vnano-integral-output/ )

        exevalator.write_variable_at(x_address, x).expect("Failed to set the value of the variable x.");
        let fx_left: f64 = exevalator.eval(expression_str).expect("Failed to evaluate the value of f(x).");

        exevalator.write_variable_at(x_address, x + delta).expect("Failed to set the value of the variable x.");
        let fx_right: f64 = exevalator.eval(expression_str).expect("Failed to evaluate the value of f(x).");

        exevalator.write_variable_at(x_address, x + delta/2.0).expect("Failed to set the value of the variable x.");
        let fx_center: f64 = exevalator.eval(expression_str).expect("Failed to evaluate the value of f(x).");

        result += (fx_left + fx_right + 4.0 * fx_center) * delta / 6.0;
    }

    // Display the result
    println!("----------");
    println!("f(x)        = {}", expression_str);
    println!("lower-limit = {}", lower_limit_value);
    println!("upper-limit = {}", upper_limit_value);
    println!("result      = {}", result);
}
